### ISSUE TITLE
Better Client-Side Library Management

### ISSUE THREAD

**Issue by darthtrevino** [atsid/generator-ats-node-webapp#108]
We need a better way of managing client-side libraries. 

Developers should use CDN-ified libraries when possible; this will let the browser asynchronously load libraries, and leave the application payload to a minimal size. For libraries that are not available via CDNs, or in use-cases where we need to bypass CDNs, we should use bower to manage those dependencies and load them into the browser directly.

**Comment by natoverse** [comment]
It would be great if this were configurable at dev time. Sometimes you may want to work offline and get them locally from pre-installed bower, but then switch back when you are online or for deployment.

**Comment by bmathews** [comment]
> we should use bower to manage those dependencies and load them into the browser directly.

Are you advocating separate requests to each dep vs concatenating down to a single vendor.js/css file?

**Comment by darthtrevino** [comment]
Yeah. Parallel requests to CDN assets seems ideal

**Comment by darthtrevino** [comment]
That's essentially what's happening now. In the Angular client, the JS bundle is just the application code.

**Comment by natoverse** [comment]
There are of course many libraries that are modularized enough, even in CDN, that it would be worthwhile to do what @bmathews is suggesting just to reduce sheer HTTP requests.

**Comment by bmathews** [comment]
Is there a number of vendor js/css files you'd reach before considering bundling them? It looks like you're at 10+, with more than 6 on a single CDN, so some of them will be deferred until others finish. Maybe something worth measuring?

Another thing worth considering is that in most cases, especially on mobile, the majority of time is spent waiting (even when in parallel) vs actually downloading. E.g., if we have 200ms latency to a server but a 20mb/s connection. 

![image](https://cloud.githubusercontent.com/assets/848347/10528720/c3b52d62-734c-11e5-80dc-53302d10b344.png)

**Comment by darthtrevino** [comment]
I'd do some research. Loading well-known libraries from CDN lets us leverage the browser cache. A vendor bundle would make sense for libraries we're loading from bower.

**Comment by natoverse** [comment]
This is getting older now, but you might be able to find newer data: http://www.stevesouders.com/blog/2013/03/18/http-archive-jquery/

**Comment by bmathews** [comment]
This builtwith site that Souders links to shows the percentage of top websites using a library hosted on a particular CDN. But even from this it's kind of hard to tell guesstimate: http://trends.builtwith.com/javascript/Google-Hosted-AngularJS
